improver/utilities/field_texture.py

```diff
@@ -0,0 +1,176 @@
+from improver.threshold import BasicThreshold
+from improver.utilities.cube_manipulation import collapsed
+
+
+class FieldTexture(BasePlugin):
+    """Plugin to calculate the texture in a binary field considering edge transitions
+
+    Code methodology:
+
+    1) Takes a binary field that has been thresholded at a cloud cover fraction
+       of interest and looks for the transitions/edges in the field that mark
+       out a transition from cloudy to non-cloudy grid cell.
+    2) The transition calculation is then fed into the neighbourhooding code
+       (_calculate_ratio) to calculate a ratio for each cell.
+    3) The new cube of ratios is then thresholded and the realization coordinates
+       are collapsed to generate a mean of the thresholded ratios.
+    """
+
+    def __init__(self, nbhood_radius, ratio_threshold):
+        """
+        The class is called with the default constraints for the processing code.
+
+        Args:
+            nbhood_radius (float):
+                A neighbourhood radius of sufficient size to capture the region
+                and all actual transitions, in metres.
+
+            ratio_threshold (float):
+                A threshold to re-normalise values about a sensible value.
+
+        """
+        self.nbhood_radius = nbhood_radius
+        self.ratio_threshold = ratio_threshold
+
+    def _calculate_ratio(self, cube, radius):
+        """
+        Calculates the ratio of each cell using neighbourhooding.
+
+        Args:
+            cube (cube):
+                Input data in cube format containing a single realization.
+
+            radius (float):
+                Average radius for neighbourhood in metres.
+
+        Returns:
+            ratio (cube):
+                A ratio between 0 and 1 of actual transitions over potential transitions.
+        """
+
+        data = self._calculate_transitions(cube.data)
+        result_cube = create_new_diagnostic_cube(
+            "cell_edge_differences", 1, cube, MANDATORY_ATTRIBUTE_DEFAULTS, data=data
+        )
+        potential_transitions = SquareNeighbourhood(sum_or_fraction="sum").run(
+            cube, radius=radius
+        )
+        potential_transitions.data = np.where(
+            cube.data > 0, potential_transitions.data * 4, 0
+        )
+
+        # Note that where there is no cloud the value is forced to 1. So high values
+        # correspond to scattered cloud or no cloud.
+
+        actual_transitions = SquareNeighbourhood(sum_or_fraction="sum").run(
+            result_cube, radius=radius
+        )
+        actual_transitions.data = np.where(cube.data > 0, actual_transitions.data, 0)
+        ratio = actual_transitions.copy(data=np.ones_like(actual_transitions.data))
+        np.divide(
+            actual_transitions.data,
+            potential_transitions.data,
+            out=ratio.data,
+            where=(potential_transitions.data > 0),
+        )
+        return ratio
+
+    def _calculate_transitions(self, data):
```

Comment:
Codacy found an issue: [Method could be a function](https://app.codacy.com/manual/metoppv_tech/improver/pullRequest?prid=5960681)